blackbird_python/blackbird/tests/test_auxiliary.py

```diff
@@ -0,0 +1,558 @@
+def parser():
+    """Returns a dummy/empty blackbird parser"""
+    return blackbirdParser("")
+
+
+@pytest.fixture
+def ctx():
+    """Returns a dummy rule context"""
+    return ParserRuleContext()
+
+
+@pytest.fixture
+def num(parser, ctx):
+    """Generates a blackbird number label"""
+    def _number_label(n):
+        """Generate the blackbird number n"""
+        class DummyNumberLabel(blackbirdParser.NumberLabelContext):
+            """Dummy class"""
+            def number(self):
+                """The number method"""
+                number = blackbirdParser.NumberContext(parser, ctx)
+                number.COMPLEX = lambda: True
+                number.getText = lambda: n
+                return number
+        return DummyNumberLabel(parser, ctx)
+    return _number_label
+
+
+@pytest.fixture
+def var(parser, ctx):
+    """Generates a blackbird variable label"""
+    def _var_label(key):
+        """Generate a reference to blackbird variable key"""
+        _var = blackbirdParser.VariableLabelContext(parser, ctx)
+        _var.getText = lambda: key
+        return _var
+    return _var_label
+
+
+class TestLiteral:
+    """Tests for the _literal function"""
+    def test_literal_string(self):
+        """Test that a Blackbird string is properly converted to a Python type"""
+        nonnumeric = blackbirdParser.NonnumericContext(parser, ctx)
+        nonnumeric.STR = lambda: True
+        nonnumeric.getText = lambda: "Test value"
+        assert _literal(nonnumeric) == "Test value"
+
+    def test_literal_bool_true(self):
+        """Test that a Blackbird bool is properly converted to a Python type"""
+        nonnumeric = blackbirdParser.NonnumericContext(parser, ctx)
+        nonnumeric.BOOL = lambda: True
+        nonnumeric.getText = lambda: "True"
+        assert _literal(nonnumeric)
+
+    def test_literal_bool_false(self):
+        """Test that a Blackbird bool is properly converted to a Python type"""
+        nonnumeric = blackbirdParser.NonnumericContext(parser, ctx)
+        nonnumeric.BOOL = lambda: True
+        nonnumeric.getText = lambda: "False"
+        assert not _literal(nonnumeric)
+
+    def test_literal_bool_invalid(self):
+        """Test that a Blackbird bool is properly converted to a Python type"""
+        nonnumeric = blackbirdParser.NonnumericContext(parser, ctx)
+        nonnumeric.BOOL = lambda: True
+        nonnumeric.getText = lambda: "false"
+
+        with pytest.raises(ValueError, match="Unknown boolean value"):
+            _literal(nonnumeric)
+
+    def test_literal_invalid(self):
+        """Test that a Blackbird bool is properly converted to a Python type"""
+        nonnumeric = blackbirdParser.NonnumericContext(parser, ctx)
+        nonnumeric.getText = lambda: ""
+
+        with pytest.raises(ValueError, match="Unknown value"):
+            _literal(nonnumeric)
+
+
+class TestNumber:
+    """Tests for the _number function"""
+
+    @pytest.mark.parametrize('n, expected', test_ints)

```

Comment:
any reason to create two different variables (`n` and `expected`) here instead of using a single one?

Reply:
Not really, more so I thought that using `n` and `expected` explicitly would be clearer to someone, instead of using `n[0]` and `n[1]`.